<h1><center>Feature Selection</center></h1>
In this Notebook,  I will use [Boruta method](https://mbq.github.io/Boruta/) to select the main features.

In [2]:
import numpy as np
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import StratifiedShuffleSplit
from sklearn.metrics import confusion_matrix
from sklearn.metrics import f1_score
from sklearn.metrics import precision_recall_fscore_support as score
from sklearn.utils import shuffle

from boruta import boruta_py

import pandas as pd
import matplotlib.pyplot as plt
import copy

from caimcaim import CAIMD

from utility import random_forest_classifer



In [3]:
data_train = pd.read_csv('data/train2.csv',index_col=0)

In [4]:
data_train_happy=data_train.loc[data_train['TARGET']==0]
data_train_unhappy=data_train.loc[data_train['TARGET']==1]
s_perc = 0.5
data_train_happy_s = data_train_happy.sample(int(s_perc*data_train_unhappy.shape[0]),random_state=21)
data_train_unhappy_s= data_train_unhappy.sample(data_train_happy_s.shape[0],random_state=32)
data_train_homo = pd.concat([data_train_happy_s,data_train_unhappy_s])

In [5]:
# define Boruta feature selection method
forest = RandomForestClassifier(n_estimators=100,max_features="sqrt",class_weight='balanced')

matrix_features = data_train_homo.as_matrix()[:,:-1]
labels = data_train_homo['TARGET'].as_matrix()


feat_selector = boruta_py.BorutaPy(forest, n_estimators='auto', verbose=2)
 
# find all relevant features
feat_selector.fit(matrix_features, labels)
 
# check selected features
features_s = feat_selector.support_
# check ranking of features
features_r = feat_selector.ranking_

Iteration: 	1 / 100
Confirmed: 	0
Tentative: 	369
Rejected: 	0
Iteration: 	2 / 100
Confirmed: 	0
Tentative: 	369
Rejected: 	0
Iteration: 	3 / 100
Confirmed: 	0
Tentative: 	369
Rejected: 	0
Iteration: 	4 / 100
Confirmed: 	0
Tentative: 	369
Rejected: 	0
Iteration: 	5 / 100
Confirmed: 	0
Tentative: 	369
Rejected: 	0
Iteration: 	6 / 100
Confirmed: 	0
Tentative: 	369
Rejected: 	0
Iteration: 	7 / 100
Confirmed: 	0
Tentative: 	369
Rejected: 	0
Iteration: 	8 / 100
Confirmed: 	0
Tentative: 	9
Rejected: 	360
Iteration: 	9 / 100
Confirmed: 	0
Tentative: 	9
Rejected: 	360
Iteration: 	10 / 100
Confirmed: 	0
Tentative: 	9
Rejected: 	360
Iteration: 	11 / 100
Confirmed: 	0
Tentative: 	9
Rejected: 	360
Iteration: 	12 / 100
Confirmed: 	0
Tentative: 	7
Rejected: 	362
Iteration: 	13 / 100
Confirmed: 	0
Tentative: 	7
Rejected: 	362
Iteration: 	14 / 100
Confirmed: 	0
Tentative: 	7
Rejected: 	362
Iteration: 	15 / 100
Confirmed: 	0
Tentative: 	7
Rejected: 	362
Iteration: 	16 / 100
Confirmed: 	0
Tentative: 	6


* BorutaPy finished running.

* Iteration: 	76 / 100
* Confirmed: 	0
* Tentative: 	0
* Rejected: 	369

From this analysis, Bouruta method is not able to find a subset of relvant features.